### DISCUSSION TITLE
tutorials/intro-typescript/05-apollo-server

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#155] [course feedback]
# tutorials/intro-typescript/05-apollo-server

https://www.apollographql.com/tutorials/intro-typescript/05-apollo-server

**Comment by pdedeyan** [comment]
As far as I can tell, I didn't miss any steps but I'm getting this error...
odyssey-intro-typescript % npm run dev

> Intro-to-GraphQL-with-TypeScript@0.0.0 dev /Users/370134/odyssey-intro-typescript
> ts-node-dev --respawn ./src/index.ts

[INFO] 09:24:12 ts-node-dev ver. 2.0.0 (using ts-node ver. 10.9.2, typescript ver. 5.5.4)
(node:23294) UnhandledPromiseRejectionWarning: Error: Query root type must be provided.
    at assertValidSchema (/Users/370134/odyssey-intro-typescript/node_modules/graphql/type/validate.js:59:11)
    at Function.generateSchemaDerivedData (/Users/370134/odyssey-intro-typescript/node_modules/@apollo/server/src/ApolloServer.ts:730:22)
    at Object.schemaDerivedDataProvider (/Users/370134/odyssey-intro-typescript/node_modules/@apollo/server/src/ApolloServer.ts:275:28)
    at new SchemaManager (/Users/370134/odyssey-intro-typescript/node_modules/@apollo/server/src/utils/schemaManager.ts:79:36)
    at new ApolloServer (/Users/370134/odyssey-intro-typescript/node_modules/@apollo/server/src/ApolloServer.ts:272:26)
    at startApolloServer (/Users/370134/odyssey-intro-typescript/src/index.ts:15:20)
    at Object.<anonymous> (/Users/370134/odyssey-intro-typescript/src/index.ts:23:3)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Module._compile (/Users/370134/odyssey-intro-typescript/node_modules/source-map-support/source-map-support.js:521:25)
    at Module.m._compile (/private/var/folders/9w/d96sf_hd4p5b22hgm39hdhyh0000gp/T/ts-node-dev-hook-26402904995380116.js:69:33)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:23294) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:23294) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Reply by pdedeyan** [reply]
Restarted and it's fine now.

**Comment by kelvenus** [comment]
Just my observation, for new coders. Put the startApolloServer() at the end of the last "}" in index.ts for your code to run properly. when you inlcude it in the function, it will not run. Ex. 
async function startApolloServer() {
  ********
 }
  startApolloServer();

intead of 

async function startApolloServer() {
  *********
  startApolloServer();
 }

I made this error.

**Reply by lizhennessy** [reply]
Thank you for calling this out @kelvenus ! Definitely an important detail to note.

**Reply by MichaelFoss** [reply]
One thing I would suggest is adding a quick blurb at the end like this:

> After all steps are finished, this is what your `index.ts` file should look like:
>
> ```typescript
> import { ApolloServer } from "@apollo/server";
> import { startStandaloneServer } from "@apollo/server/standalone";
>
> import { readFileSync } from "fs";
> import path from "path";
> import { gql } from "graphql-tag";
> 
> async function startApolloServer() {
>  const server = new ApolloServer({ typeDefs });
>  const { url } = await startStandaloneServer(server);
>  console.log(`
>    🚀  Server is running!
>    📭  Query at ${url}
>  `);
> }
> 
> startApolloServer();
> ```

This provides anyone who incorrectly implemented a step a way to see where things went wrong (the above comment being an example!).

**Reply by MichaelFoss** [reply]
Oh, and of course include the `typeDefs` definition:

```typescript
const typeDefs = gql(
  readFileSync(path.resolve(__dirname, "./schema.graphql"), {
    encoding: "utf-8",
  })
);
```

**Reply by kelvenus** [reply]
Thanks, I will update mine.  I appreciate you adding additional tips for us @MichaelFoss

**Comment by DaleSeo** [comment]
The instructions for this lesson seems a bit outdated compared to [the getting started docs](https://www.apollographql.com/docs/apollo-server/getting-started), where top-level `await` is used and `#graphql` is used instead of the `gql` utility. I'd like to suggest keeping the content in sync to avoid confusion.

**Reply by lizhennessy** [reply]
Thanks for calling this out @DaleSeo ! We'll take a look at this. There may be some subtle differences regarding the JavaScript and TypeScript implementations as described by the docs, but we'll make sure all the latest features are reflected here.

**Reply by lizhennessy** [reply]
Ah I see what you mean now. The quickstart guide shows how we can get a server instance up and running very quickly with type definitions and resolvers in the same file. We can definitely do it this way; but because our types are in a separate `schema.graphql` file, we're using the `gql` utility to prepare them in a format that the server can more readily digest. The end effect is the same, these are two different ways of going about it. With regards to the top-level `await`, please feel free to use this format too! We've chosen to encapsulate the server bootup code in a single function that the file calls, as this gives us some portability in case we want to compose the function with other startup requirements. But both approaches work!

**Reply by DaleSeo** [reply]
@lizhennessy Thanks for taking the time to elaborate on the approaches. It maks a lot of sense now!

**Comment by HewaliGogoi** [comment]
for someone who want to use ES6 module, the code looks something like this:
``` 
import { ApolloServer } from "@apollo/server";
import { startStandaloneServer } from "@apollo/server/standalone";
import { readFileSync } from "fs";
import { gql } from "graphql-tag";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const typeDefs = gql(
  readFileSync(path.resolve(__dirname, "./schema.graphql"), {
    encoding: "utf-8",
  })
);

async function startApolloServer() {
  const server = new ApolloServer({ typeDefs });
  const { url } = await startStandaloneServer(server);

  console.log(`
    🚀  Server is running!
    📭  Query at ${url}
  `);
}

startApolloServer(); 
```

However, I've not use the top-level ``await`` as the single function gives me a better picture.
